Add Appveyor CI for Windows

zbjornson opened the pull request:
This mostly works. VS2015 fails without #670. We need #458 merged for any to fully pass -- see log:
https://ci.appveyor.com/project/zbjornson/node-canvas/build/job/x3domgfjsq77445m After that I need to change the script to install libjpeg-turbo.

For now there's only one node env. Think it's unnecessary to have more than one, but it's easy to add.

(Sorry I've got 83a8b27 upstream from #671 in here.)

LinusU commented on appveyor.yml:
```diff
@@ -0,0 +1,17 @@
+  - npm install
+build: off
+test_script:
+- cmd: mocha test/*.test.js
```
Newline at the end of the file please :)

LinusU replied:
Also, shouldn't the `cmd` be `npm test`?

zbjornson replied:
`npm test` [runs pretest, then test](https://docs.npmjs.com/misc/scripts). pretest currently builds. `npm install` also builds, so it was doing a redundant build. Do we need pretest to build? I agree it would be nicer to just have `npm test`.

LinusU replied:
Ahh, hmm, I would kind of like to get rid of the `pre*` as well. I don't think it's the right place to have that responsibility here.

Although, at least on unix, `node-gyp build` isn't the same as **`re`**`build`. Mine just outputs "nothing to be done"...

LinusU commented:
This is awesome, some comments:

- can you rebase on master now that I have pulled in the first commit?
- can we use [matrix building](http://www.appveyor.com/docs/lang/nodejs-iojs#testing-under-multiple-versions-of-node-js-or-io-js) to test on multiple versions of Node.js?
- is there some type of account on Appveyor?

zbjornson commented:
Added the full matrix that Travis is set to use. Aside from the jpeg-stream test, all versions except 1.8 pass. 

I think 1.8 fails because of [some race condition](https://github.com/npm/npm/issues/8687) that's resolved around version 2.3. Commented that version out.

Changed the test_script to `npm test` assuming #676 will land.

Logs are a bit noisy, pending https://github.com/npm/npm/issues/10511.

https://ci.appveyor.com/project/zbjornson/node-canvas

@LinusU you will need to make a free Appveyor account and enable a build for this repo (otherwise it will only run on my fork). Then it'll be like Travis and run on pushes. Check out the settings tab in Appveyor for badges, etc. There's a check box to skip branches without appveyor.yml, so you could set this up before this PR lands.

[![Build status](https://ci.appveyor.com/api/projects/status/1dx9ramd23ptl8w0?svg=true)](https://ci.appveyor.com/project/zbjornson/node-canvas)

LinusU commented on appveyor.yml:
```diff
@@ -0,0 +1,34 @@
+  - npm install
+build: off
+test_script:
+- cmd: npm test
```
indentation?

LinusU commented on appveyor.yml:
```diff
@@ -0,0 +1,34 @@
+  - 7z x gtk.zip -oGTK > nul
+  - mv GTK/ C:/
+  - npm install
+build: off
```
Would it work if removed both of these lines?

```yml
  - npm install
build: off
```

zbjornson replied:
No, the default build command invokes msbuild directly. Getting the configuration right for that looks more verbose/complicated than it currently is.

LinusU replied:
Cool, just checking :)

LinusU commented:
Do you think we should remove all iojs releases from both here and travis? I don't know what the LTS status is on those but I don't think anyone uses them anymore...

zbjornson commented:
From https://github.com/nodejs/LTS/

> There will be no LTS releases cut from the nodejs/io.js stream.

I don't know if there are any metrics available to see if anyone is using iojs nonetheless. Doesn't hurt to leave them, I think?

LinusU commented:
@TooTallNate @domenic @rvagg @kangax Who should I talk to to get a free appveyor account set up for Automattic? I guess it would be good if someone with a somewhat official email address set it up? If this is desirable at all, which I assume it is :)

kangax commented:
That would be @TooTallNate or @rvagg (@domenic and I were invited)

rvagg commented:
Not I, it doesn't show up for me in appveyor so I guess I don't have enough access in the Automattic org, will have to defer to someone more official, @TooTallNate or someone else there.

LinusU commented:
@TooTallNate Do you have time to take a look at this? I would appreciate it!

chearon commented:
This would be nice to have, I don't use node-canvas on Windows anymore, so AppVeyor would have caught #984 breaking the build against the old gtk.zip

LinusU commented:
I'm personally still open for this. One thing I don't like with AppVeyor though is that they are super super super slow. If anyone knows any other free for open source CI for Windows, I would be very interested 😄 

Other than that, I think we are at the same position as before, someone from the automattic org needs to enable this...

chearon commented:
> One thing I don't like with AppVeyor though is that they are super super super slow

Yeah, so basically the same problem as #796. I wonder how much it would help if we only built with the latest node on macOS and Windows? I could maybe experiment with that a bit.

zbjornson commented:
Just updated this branch since JPEG works on Windows now.

The builds themselves aren't slow (they're actually about a minute faster than Travis) but they all run sequentially. So, I've set the matrix to just Node 6 (LTS) and MSVS 2013 and 2015. (2015 is probably sufficient; 2017 I haven't got working yet, although I use it on one of my computers.)

✔️ https://ci.appveyor.com/project/zbjornson/node-canvas/build/0.0.61 <-- PR #987

LinusU commented:
What I remember with them being super slow was for them to actually start, my latest build in node-spearker took hours (I think, can't find any numbers now after the fact) just to start 😱 

Anyhow, I'm still all for getting this in, just venting my frustation 😂

TooTallNate commented:
I no longer have admin access to this repo either. Not sure who at a8c to ping based on the commit logs.

Maybe /cc @dmsnell? I dunno.

dmsnell commented:
Thanks @TooTallNate! I can try and look into this soon.

zbjornson commented:
Hey @dmsnell, any luck finding the admin of this repo? :)

dmsnell commented:
@zbjornson do we need the fix in canvas.cc in order to add Appveyor? Can you provide some context for that fix and why it's there, what it's fixing? Seems like we're changing the line under an existing comment and it's not clear if that comment needs to be altered before merging this in.

zbjornson commented:
That commit was already merged, I just haven't rebased. We need the admin of this repo to enable Appveyor though.

dmsnell commented:
@zbjornson I'm getting the Appveyor account setup. I may need some help in the next few days getting it configured as we need it to be here, but I'll let you know.

dmsnell commented:
@zbjornson in the meantime, would you mind rebasing this branch against the current master?

zbjornson commented:
@dmsnell great to hear, thank! Rebased.

(While I have your attention... https://github.com/Automattic/node-canvas/issues/914 is the other issue where we're seeking the help of an admin.)

dmsnell commented:
Although the Appveyor account isn't established yet I'm going to go ahead and merge this in for now. Honestly I don't know how to review this but I can tell it shouldn't cause trouble. Thanks for your contributions!
